### PR TITLE
Implement daily worked time rounding

### LINKED ISSUE #147: [Feature request] Rounding of time entries
I track my working time only with a 15 minute precision, i.e. only with the granularity of a quarter hour. Others might use 5 minutes as the smallest chunk of work time. Currently doing so is quite tedious since when I do not hit start or stop exactly at a multiple of 15 minutes, I have to correct the entry manually afterwards to make it fit that granularity.

Therefore it would be great when there was a configuration option to automatically round the time entries to the next full 15 minutes (or 5 minutes or whatever someone else uses as their minimal time chunk).

### PR DESCRIPTION
**Pull request by sant0s12**
Attempts to resolve #147.

At the moment it does not work when generating reports.

### CODE REVIEW

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculatorV2.java [hunk 1]
```diff
@@ -138,17 +144,24 @@ public boolean containsEvents() {
...
-		this.handleFlexiTime = handleFlexiTime;
+		this.handleFlexiTime = preferences.getBoolean(Key.ENABLE_FLEXI_TIME.getName(), false);
 		if (handleFlexiTime) {
 			this.flexiReset = timerManager.getFlexiReset();
 		} else {
 			this.flexiReset = FlexiReset.NONE;
 		}
 
+		this.roundingEnabled = preferences.getBoolean(Key.ROUNDING_ENABLED.getName(), false);
+		if (preferences.getBoolean(Key.ROUNDING_ENABLED.getName(), false)) {
```
Comment (mathisdt) [inline app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculatorV2.java:159]: This can be
```
if (roundingEnabled) {
```
instead (we just evaluated the same statement used here).

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculatorV2.java [hunk 2]
```diff
@@ -138,17 +144,24 @@ public boolean containsEvents() {
...
 
-		this.handleFlexiTime = handleFlexiTime;
+		this.handleFlexiTime = preferences.getBoolean(Key.ENABLE_FLEXI_TIME.getName(), false);
 		if (handleFlexiTime) {
 			this.flexiReset = timerManager.getFlexiReset();
 		} else {
 			this.flexiReset = FlexiReset.NONE;
 		}
 
+		this.roundingEnabled = preferences.getBoolean(Key.ROUNDING_ENABLED.getName(), false);
```
Comment (mathisdt) [inline app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculatorV2.java:158]: `this.` is superfluous here (above it is necessary because the parameter has the same name as the instance variable).

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculatorV2.java [hunk 3]
```diff
@@ -553,4 +566,12 @@ public void calculatePeriod(PeriodEnum period, boolean includeFlexiTime) {
 		this.currentDayActual = currentDayActual;
 		this.currentDayTarget = currentDayTarget;
 	}
+
+	public static long roundUpToMultiple(long value, long multiple) {
+		if (multiple == 1) {
+			return value;
+		} else {
+			return (value + multiple - 1) / multiple * multiple;
```
Comment (mathisdt) [inline app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculatorV2.java:574]: This won't work in all cases. Here's an example calculation for the inputs **value = 43, multiple = 15** which should round to 45.

But it doesn't, the method would return 57 with these inputs.

To be on the safe side, I'd recommend creating a unit test for the calulation method where some "normal" cases and many (if not all) of the edge cases are checked. For a start you can copy `TimeCalculatorTest` to `TimeCalculatorV2Test` and adjust the content accordingly.

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculatorV2.java [hunk 4]
```diff
@@ -361,7 +374,7 @@ public void calculateNextDay() {
 
 		currentDayHasEvents = (events != null && !events.isEmpty());
 
-		workedTime += calculateWorkTime(events);
+		workedTime += roundUpToMultiple(calculateWorkTime(events), smallestTimeUnit);
```
Comment (mathisdt) [inline app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculatorV2.java:377]: This should be wrapped in an `if` block to only execute it when required:
```
if (roundingEnabled) {
   workedTime += roundUpToMultiple(calculateWorkTime(events), smallestTimeUnit);
}
```

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculatorV2.java [hunk 5]
```diff
@@ -138,17 +144,24 @@ public boolean containsEvents() {
...
 			this.flexiReset = timerManager.getFlexiReset();
 		} else {
 			this.flexiReset = FlexiReset.NONE;
 		}
 
+		this.roundingEnabled = preferences.getBoolean(Key.ROUNDING_ENABLED.getName(), false);
+		if (preferences.getBoolean(Key.ROUNDING_ENABLED.getName(), false)) {
+			this.smallestTimeUnit = Integer.parseInt(preferences.getString(Key.SMALLEST_TIME_UNIT.getName(), "1"));
+		} else {
+			this.smallestTimeUnit = 1;
```
Comment (mathisdt) [inline app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculatorV2.java:162]: The `else` part can be left out if the variable is initialized with the value `1` above:
```
private final int smallestTimeUnit = 1;
```

**Review by mathisdt — CHANGES_REQUESTED** [review]
Hi @sant0s12, 
thanks for your pull request! I'm sorry it took me so long to review it - I just started into a new job which took some adjusting, and then I was away on family vacation. But now I provided some feedback (see code comments).

For your feature to be complete, it will be necessary to also **include rounding for the reports** if it's activated. The users won't tolerate different numbers in the app and in the reports (there already were some issues reported regarding this in the past).

But more importantly the **rounding formula** seems to cover only some cases. As I wrote in the code comment I'd like to have a unit test for this calculation. If you are unsure how to approach this, I can also help out with that, but at the moment the rounding seems not to work correctly. So if you don't add a unit test, please at least fix the method `roundUpToMultiple()`.

Thanks again for your contribution! It didn't deserve to be ignored for so long!

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculatorV2.java (hunk 3 above)
Comment (mathisdt) [reply app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculatorV2.java:574]: Possible values of `multiple` are: 1, 2, 3, 4, 5, 6, 10, 12, 15, 20, 30, 60

But probably only the values 5, 10, 15, 30 and 60 will be used in real life...

**Comment by sant0s12** [comment]
Hi @mathisdt, thanks a lot for your feedback. I'll take a look at it and implement your suggestions.

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculatorV2.java (hunk 3 above)
Comment (sant0s12) [reply app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculatorV2.java:574]: Oh, that's weird. I'll write some tests for it.

Also, I do use 6 as a multiple myself since my work tracks time to an accuracy of 0.1 hours. Therefore I think it makes sense to leave it up to the user to choose any divisor of 60 :)

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculatorV2.java (hunk 5 above)
Comment (sant0s12) [reply app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculatorV2.java:162]: Maybe I am missing something, but I get an error if I do that.

Comment (mathisdt) [reply app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculatorV2.java:162]: Sorry, my bad - the `final` keyword is the problem. With
```
private int smallestTimeUnit = 1;
```
it works.

**Comment by Josic490** [comment]
Hi @sant0s12,

I'm interested in seeing this feature deployed.
Are you still working on it?

File: app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculatorV2.java (hunk 4 above)
Comment (mathisdt) [reply app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculatorV2.java:377]: While glancing over this again, my previous comment doesn't make sense. The `calculateWorkTime(events)` would also be only executed if my code snippet was used, so it better should be like:

```
if (roundingEnabled) {
	workedTime += roundUpToMultiple(calculateWorkTime(events), smallestTimeUnit);
} else {
	workedTime += calculateWorkTime(events);
}
```
